src/ocean/text/convert/Formatter.d

```diff
@@ -478,7 +478,29 @@ private void handle (T) (T v, FormatInfo f, scope FormatterSink sf, scope ElemSi
         nullWrapper(&v, se(v.toString(), f), se("null", f));
     else static if (is(T == interface))
         handle!(Object)(cast(Object) v, f, sf, se);
+    else static if (isInputRange!T)
+    {
+        if (v.empty)
+        {
+            sf("[]");
+            return;
+        }
 
+        Flags old_flag = f.flags;
+        f.flags |= Flags.Nested;
+        sf("[ ");
+        while (1)
+        {
+            handle(v.front, f, sf, se);
+            v.popFront();
+            if (v.empty)
+                break;
+            else
+                sf(", ");
+        }
+        sf(" ]");
```

Comment:
```suggestion
        sf("[");
        while (!v.empty)
        {
            handle(v.front, f, sf, se);
            v.popFront();
            if (!v.empty)
                sf(", ");
        }
        sf("]");
```

And you can remove the special case for empty ranges

Reply:
Oh, great. It's very simple, it's clear